File: php/src/Advertisements/Advertisement/Application/Command/RenewAdvertisement/RenewAdvertisementUseCase.php
<?php
declare(strict_types=1);

namespace Demo\App\Advertisements\Advertisement\Application\Command\RenewAdvertisement;

use Demo\App\Advertisements\Advertisement\Application\Exceptions\InvalidPasswordException;
use Demo\App\Advertisements\Advertisement\Domain\AdvertisementRepository;
use Demo\App\Advertisements\Advertisement\Domain\ValueObjects\AdvertisementId;
use Demo\App\Advertisements\Shared\ValueObjects\Password;
use Exception;

final class RenewAdvertisementUseCase
{
    public function __construct(private AdvertisementRepository $advertisementRepository)
    {
    }

    /**
     * @throws Exception
     */
    public function execute(RenewAdvertisementCommand $command): void
    {
        // TODO: Implement user security

        $advertisement = $this->advertisementRepository->findByIdOrFail(new AdvertisementId($command->id));

        if (!$advertisement->password()->isValidatedWith($command->password)) {
            throw InvalidPasswordException::build();
        }

        $advertisement->renew(Password::fromPlainPassword($command->password));

        $this->advertisementRepository->save($advertisement);
    }
}

File: kotlin/src/main/kotlin/advertisements/advertisement/application/command/renewadvertisement/RenewAdvertisementUseCase.kt
package advertisements.advertisement.application.command.renewadvertisement

import advertisements.advertisement.domain.exceptions.AdvertisementNotFoundException
import advertisements.advertisement.application.exceptions.PasswordDoesNotMatchException
import advertisements.advertisement.domain.AdvertisementRepository
import advertisements.advertisement.domain.value_object.AdvertisementId
import advertisements.shared.value_object.Password
import framework.database.TransactionManager

class RenewAdvertisementUseCase(
    private val advertisementRepository: AdvertisementRepository,
    private val transactionManager: TransactionManager,
) {
    fun execute(renewAdvertisementCommand: RenewAdvertisementCommand) {
        transactionManager.beginTransaction()

        try {
            val advertisementId = AdvertisementId(renewAdvertisementCommand.id)
            val advertisement = advertisementRepository.findById(advertisementId)

            if (null === advertisement) {
                throw AdvertisementNotFoundException.withId(advertisementId.value())
            }

            if (!advertisement.password.isValidatedWith(renewAdvertisementCommand.password))
                throw PasswordDoesNotMatchException.build()

            advertisement.renew(Password.fromPlainPassword(renewAdvertisementCommand.password))

            advertisementRepository.save(advertisement)
            transactionManager.commit()
        } catch (e: Exception) {
            transactionManager.rollback()
            throw e
        }
    }
}
